### PR TITLE
Traffic rate controller

### PR DESCRIPTION
**Pull request by Garnek20**
- extracted logic of traffic control from scope to `TrafficRateController`
- added limiter for http query param `callee_id`
- added `dry_mode` to rate limiter

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -272,6 +272,12 @@ func applyConfig(cfg *config.Config) error {
 		return err
 	}
 
+	controller, err := NewTrafficRateController(cfg.RateLimiter, redisProxy)
+
+	if err != nil {
+		return err
```
Comment (r-zenine) [inline main.go:278]: Can you enrich the error : 
```go
return fmt.Errorf("while doing ....., got error %w")
```

File: proxy_test.go [hunk 2]
```diff
@@ -62,6 +62,8 @@ func newConfiguredProxy(cfg *config.Config) (*reverseProxy, *scopeManager, error
 		return nil, nil, fmt.Errorf("error while loading config: %s", err)
 	}
 	p.withScopeManager(scManager)
+	noLimiterController, _ := NewTrafficRateController(config.RateLimiterConfig{Limit: 0}, nil)
```
Comment (r-zenine) [inline proxy_test.go:65]: You might want to check the error `_` and `Fatal` if it's not nil. 
It might save hours of debug in a year if that test fails for an unexpected reason

File: scope.go [hunk 3]
```diff
@@ -45,6 +45,7 @@ type scope struct {
 
 	sessionId      string
 	sessionTimeout int
+	calleId        string
```
Comment (r-zenine) [inline scope.go:48]: is it `callee` ?

Comment (Garnek20) [reply scope.go:48]: a typo, it should be `calleeId`

File: traffic_rate_controller.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+type trafficRateController struct {
+	ratelimiter.RateLimiter
+	windowDuration time.Duration
+	limit          uint32
+	dryRateLimit   bool
+}
+
+func (r *trafficRateController) AllowRequest(s *scope) *SlowDownErr {
```
Comment (r-zenine) [inline traffic_rate_controller.go:46]: This is hard to follow, 
If you could refactor it a bit, it would help me review it :)

File: traffic_rate_controller.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+func (r *trafficRateController) allow(s *scope, labels prometheus.Labels) *SlowDownErr {
+	uQueries := s.user.queryCounter.Inc()
+	cQueries := s.clusterUser.queryCounter.Inc()
+
+	decUserCnt := func() { s.user.queryCounter.Dec() }
+	decClusterCnt := func() { s.clusterUser.queryCounter.Dec() }
+
+	var err error
+	if s.user.maxConcurrentQueries > 0 && uQueries > s.user.maxConcurrentQueries {
```
Comment (r-zenine) [inline traffic_rate_controller.go:155]: can you get each of these in a dedicated function with a meaningful name. it help help outline the business logic of the function 
```go
if hasUserExceededHisLimits ( uQueries, s.user.MaxConcurrentQueries ) { 
```
```

File: traffic_rate_controller.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+		err = fmt.Errorf("limits for user %q are exceeded: max_concurrent_queries limit: %d",
+			s.user.name, s.user.maxConcurrentQueries)
+	}
+	if s.clusterUser.maxConcurrentQueries > 0 && cQueries > s.clusterUser.maxConcurrentQueries {
```
Comment (r-zenine) [inline traffic_rate_controller.go:159]:
```go
if areClusterLimitsExceeded ( cQueries, s.clusterUser.MaxConcurrentQueries ) 
```

File: traffic_rate_controller.go [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+		decClusterCnt()
+		return &SlowDownErr{err: err}
+	}
+
+	s.host.Inc()
+	concurrentQueries.With(s.labels).Inc()
+	return nil
+}
+
+func (r *trafficRateController) rateLimitUser(s *scope, labels prometheus.Labels) error {
```
Comment (r-zenine) [inline traffic_rate_controller.go:187]: The 3 functions below looks almost exactly the same ? 
Is there a way to refactor this and eliminate the code duplicate ?

**Review by r-zenine — COMMENTED** [review]
I suggested some changes, 
There are some function I could not review , I left a comment there

File: traffic_rate_controller.go (hunk 7 above)
Comment (Garnek20) [reply traffic_rate_controller.go]: Done

File: traffic_rate_controller.go (hunk 4 above)
Comment (Garnek20) [reply traffic_rate_controller.go:35]: I did my best ;)
